File: matrix_bfs_dfs/2101.detonate-the-maximum-bombs.cpp
/*
 * @lc app=leetcode id=2101 lang=cpp
 *
 * [2101] Detonate the Maximum Bombs
 */

#include "algorithm"
#include "iostream"
#include "map"
#include "numeric"
#include "queue"
#include "set"
#include "vector"

using namespace std;

// @lc code=start
using LL = long long;
class Solution {
  vector<int> next[100];

public:
  int maximumDetonation(vector<vector<int>> &bombs) {
    int n = bombs.size();
    int cnt = 0;
    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < n; ++j) {
        LL dx = bombs[i][0] - bombs[j][0];
        LL dy = bombs[i][1] - bombs[j][1];
        LL dr = bombs[i][2];
        if (dx * dx + dy * dy <= dr * dr) {
          next[i].push_back(j);
        }
      }
    }
    for (size_t s = 0; s < n; ++s) {
      queue<int> q;
      q.push(s);

      vector<bool> visited(n, false);
      visited[s] = true;

      while (!q.empty()) {
        int curr = q.front();
        q.pop();
        for (auto val : next[curr]) {
          if (visited[val])
            continue;
          q.push(val);
          visited[val] = true;
        }
      }
      int count = 0;
      for (size_t i = 0; i < n; ++i) {
        count += visited[i];
      }

      cnt = max(cnt, count);
    }
    return cnt;
  }
};


// @lc code=end

File: matrix_bfs_dfs/2101.detonate-the-maximum-bombs.py
#
# @lc app=leetcode id=2101 lang=python3
#
# [2101] Detonate the Maximum Bombs
#

# @lc code=start
from pprint import pprint
from typing import List


class Solution:
    def maximumDetonation(self, bombs: List[List[int]]) -> int:
        cnt, n = 0, len(bombs)
        ori = set(range(n))
        dic = {}
        for i in range(n):
            dic[i] = set()

        for i in range(n):
            [x1, y1, r1] = bombs[i]
            for j in range(i+1, n):
                [x2, y2, r2] = bombs[j]

                sqare_distance = pow(x1-x2, 2)+pow(y1-y2, 2)
                if sqare_distance <= pow(r1, 2):
                    dic[i].add(j)
                if sqare_distance <= pow(r2, 2):
                    dic[j].add(i)

        while ori:
            i = ori.pop()
            cur = diff = dic[i]
            cur.add(i)
            pre = cur

            while diff:
                for next in diff:
                    cur = cur | dic[next]
                    # cur = set.union(cur,dic[next])

                diff = cur-pre
                pre = cur

            ori -= pre
            cnt = max(cnt, len(cur))

        return cnt


# @lc code=end
ori = set(range(5))
ori2 = set(range(5, 10))

print(ori | ori2)

# x1, y1, r1 = [10, 10, 5]
# print(x1)
# print(y1)
# print(r1)

s = Solution()
print(s.maximumDetonation(
    [[1, 2, 3], [2, 3, 1], [3, 4, 2], [4, 5, 3], [5, 6, 4]]))
